pkg/controller/common/util.go

```diff
@@ -0,0 +1,60 @@
+package common
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	v1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+// ConstructRevisionName will generate revisionName from componentName
+// will be <componentName>-v<RevisionNumber>, for example: comp-v1
+func ConstructRevisionName(componentName string, revision int64) string {
+	return strings.Join([]string{componentName, fmt.Sprintf("v%d", revision)}, "-")
+}
+
+// ExtractComponentName will extract componentName from revisionName
+func ExtractComponentName(revisionName string) string {
+	splits := strings.Split(revisionName, "-")
+	return strings.Join(splits[0:len(splits)-1], "-")
+}
+
+// CompareWithRevision compares a component's spec with the component's latest revision content
+func CompareWithRevision(ctx context.Context, c client.Client, logger logging.Logger, componentName, nameSpace,
+	latestRevision string, curCompSpec *v1alpha2.ComponentSpec) (bool, error) {
+	oldRev := &v1.ControllerRevision{}
+	// retry on NotFound since we update the component last revision first
```

Comment:
this is related to the order of the "update component status" before "create a new revision" change which is the right way to update.  In that case, Not found is expected if the component's status revision is ahead of the real revision creation. 

 I reverted that change since it broke too many tests.   The old order can lead to permanent new revision creation failure unless we handle the "already created" error with special logic.